src/phoenix/core/model_schema.py

```diff
@@ -0,0 +1,1067 @@
+
+    def __call__(
+        self,
+        *dataframes: Union[pd.DataFrame, Tuple[Name, pd.DataFrame]],
+        **kwargs: Any,
+    ) -> Model:
+        """Dimensions are the "baton" that Schema hands over to Model."""
+        _raise_if_too_many_dataframes(len(dataframes))
+        return Model(iter(self._dimensions), dataframes, **kwargs)
+
+    def to_json(self, **kwargs: Any) -> str:
+        return json.dumps(_jsonify(self), **kwargs)
+
+    @classmethod
+    def from_json(cls, json_string: str) -> Any:
+        json_data = json.loads(json_string)
+        return cls(**_objectify(json_data))
+
+
+def _agg_min_max(series: "pd.Series[Any]") -> "pd.Series[Any]":
+    return series.agg(["min", "max"])
+
+
+def _get_omitted_column_names(
+    dimensions: Iterable[Dimension],
+    dataframes: Iterable[pd.DataFrame],
+) -> Iterator[str]:
+    dataframe_columns = chain.from_iterable(df.columns for df in dataframes)
+    schema_columns = chain.from_iterable(dimensions)
+    yield from set(dataframe_columns) - set(schema_columns)
+
+
+def _group_names_by_dim_role(
+    dimensions: Iterable[Dimension],
+) -> Iterator[Tuple[DimensionRole, List[str]]]:
+    for role, dims in groupby(sorted(dimensions, key=_dim_role), key=_dim_role):
+        yield role, list(map(str, dims))
+
+
+def _guess_data_type(series: Iterable["pd.Series[Any]"]) -> DataType:
+    for s in series:
+        if is_numeric_dtype(s) or is_datetime64_any_dtype(s):
+            return CONTINUOUS
+    return DISCRETE
+
+
+_H = TypeVar("_H", bound=Hashable)
+
+
+def _dedup(it: Iterable[_H]) -> Iterator[_H]:
+    seen: Set[_H] = set()
+    for item in it:
+        if item not in seen:
+            seen.add(item)
+            yield item
+
+
+@lru_cache(maxsize=len(DatasetRole))
+# Note that NaN can't be cached, because NaN != NaN.
+def _series_constant(length: int, constant: Any) -> "pd.Series[Any]":
+    return pd.Series(np.full(length, constant))
+
+
+def _series_uuid(length: int) -> "pd.Series[str]":
+    return pd.Series(map(lambda _: uuid4(), range(length)))
+
+
+def _raise_if_too_many_dataframes(given: int) -> None:
+    if not 0 < given <= (limit := len(DatasetRole)):
+        raise ValueError("expected between 1 to %s dataframes, but %s were given" % (limit, given))
+
+
+def _ensure_str(obj: Optional[str]) -> str:
+    return "" if obj is None else str(obj)
+
+
+def _ensure_tuple(
+    dataframes: Iterable[Union[pd.DataFrame, Tuple[Name, pd.DataFrame]]],
+) -> Iterator[Tuple[Name, pd.DataFrame]]:
+    for df in dataframes:
+        if type(df) is pd.DataFrame:
+            yield (_rand_str(), df)
+        elif _is_named_df(df):
+            yield df
+        else:
+            raise ValueError("unexpected type: %s" % type(df))
+
+
+def _ensure_str_column_names(
+    dataframes: Iterable[pd.DataFrame],
+) -> Iterator[pd.DataFrame]:
+    for df in dataframes:
+        yield df.set_axis(df.columns.astype(str), axis=1)
+
+
+_T = TypeVar("_T")
+
```

Comment:
Same here. Naming is hard but we have to do better than T

Reply:
The privatizing of the T is still pretty odd syntax. I'm thinking it's hindering the readability quite a bit - I think we should consider not doing this long term unless it's paramount for it to be private.